sygnal/apnspushkin.py

```diff
@@ -455,3 +490,101 @@ async def _send_notification(self, request):
         return await Deferred.fromFuture(
             asyncio.ensure_future(self.apns_client.send_notification(request))
         )
+
+
+class ProxyingEventLoopWrapper:
+    """
+    This is a wrapper for an asyncio.AbstractEventLoop which intercepts calls to
+    create_connection and transparently tunnels them through an HTTP CONNECT
+    proxy.
+
+    Note that this may only work against aioapns and perhaps not in a wider
+    setting.
+    """
+
+    def __init__(
+        self,
+        wrapped_loop: asyncio.AbstractEventLoop,
+        proxy_address: str,
+        proxy_basic_auth: Optional[Tuple[str, str]],
+    ):
+        """
+        Args:
+            wrapped_loop:
+                the underlying Event Loop to wrap
+            proxy_address (str):
+                The address of the HTTP proxy to use.
+                Used to connect to the proxy, as well as in
+                the `Host` request header to the proxy.
+                Examples: '127.0.3.200:8080' or `prox:8080`
+            proxy_basic_auth ((str, str) or None):
+                Pass a pair of (username, password) credentials if your HTTP
+                proxy requires Proxy Basic Authentication (using a
+                Proxy-Authorization: basic ... header).
+        """
+        self._wrapped_loop = wrapped_loop
+        self.proxy_address = proxy_address
+        self.proxy_basic_auth = proxy_basic_auth
+
+    async def create_connection(
+        self,
+        protocol_factory: Callable[[], asyncio.Protocol],
+        host: str,
+        port: int,
+        ssl=False,
+    ):
+        proxy_url = urlparse(self.proxy_address)
+
+        # MUST await on this to receive TLS alerts
+        # in the event of no TLS, it will be marked as complete unconditionally
+        tls_waiter = Future()
+
+        def make_protocol():
+            if ssl:
+
+                def make_ssl():
+                    context = None
+                    if isinstance(ssl, SSLContext):
+                        context = ssl
+
+                    top_protocol = protocol_factory()
+                    ssl_protocol = SSLProtocol(
+                        loop=self._wrapped_loop,
+                        app_protocol=top_protocol,
+                        sslcontext=context,
+                        waiter=tls_waiter,
+                        server_side=False,
+                        server_hostname=proxy_url.hostname,
+                    )
+                    return ssl_protocol, top_protocol
```

Comment:
this seems to use private bits of the asyncio package...

aren't you supposed to use https://docs.python.org/3/library/asyncio-eventloop.html#tls-upgrade?

Reply:
Whoops, I didn't realise it was private! I'll change it.